aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -2482,12 +2481,23 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, const op
     // 'ord' is int or None
     std::vector<int64_t> dim_ = opt_dim.has_value() ? opt_dim.value().vec() : make_dim_list(ndim);
     if (!opt_num_ord.has_value() || dim_.size() == 1) {
-      Tensor result_ = at::linalg_vector_norm(
-          self, opt_num_ord.value_or(2), opt_dim, keepdim, opt_dtype);
-      // TODO: Resize and copy should be avoided with
-      //       https://github.com/pytorch/pytorch/issues/52712
-      at::native::resize_output(result, result_.sizes());
-      result.copy_(result_);
+      // The linalg_vector_norm() function must ensure that the result is the same dtype as
+      // opt_dtype. This check is done so that the _out variant can be called and copying
+      // can be avoided in case the opt_dtype is not supplied. linalg_vector_norm() also
+      // handles complex tensors separately using abs() and hence we use isComplexType().
+      if (isComplexType(self.scalar_type()) ||
```

Comment:
Why doesn't it work for complex dtypes?
Could you file an issue noting all these problems with `linalg_vector_norm`?

Reply:
The `linalg_vector_norm` calls `abs()` on complex tensors for making sure that autograd produces the same result after the call to `abs()` in autograd. To do this, it will cast the `self` to the `in_dtype` and call `abs()`. If you pass the `result` tensor from the `out=` parameter into the `linalg_vector_norm_out` where the result is not already a complex tensor, the function will throw an error.

A potential fix is probably to cast the `result` tensor passed from `out=` to whatever type is necessary. However I am not sure if this is coherent with the rest of pytorch. 

See this:  https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/LinearAlgebra.cpp#L2559

Reply:
If I understand your analysis correctly I think you are correct, @v0dro. See https://github.com/pytorch/pytorch/wiki/Developer-FAQ#how-does-out-work-in-pytorch. The `out=` tensor must (conceptually) be copied to, so its dtype cannot be modified.

Reply:
I think that there's two different issues here. So, on the one hand, we have that this function (for complex inputs) and `linalg_vector_norm` are functions from C to R (complex to real). As such, the out for **both** of them (in particular the `Tensor result_` returned by `at::linalg_vector_norm` two lines below) should be complex tensors. Hence, I don't understand why do we need the copy.

Now, the other thing to note here is that the 2-norm may be implemented as a complex pointwise operation (`mul`) + a complex reduction (`sum`) + a map to the reals (`abs`). There, it is true that we need an auxiliary tensor to hold the result of the first two operations. This auxiliary tensor should be complex. But then, I would assume that the implementation (if it's not a composite function) would use `abs_out` to copy the final result into the real tensor, and then it would return the `result` tensor. Now, note that this copy into a complex tensor should happen inside of `linalg_vector_norm_out`, so I still don't understand why this copy is necessary here.